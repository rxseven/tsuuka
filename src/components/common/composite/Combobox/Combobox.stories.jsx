import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

// import Dropdown from 'components/common/modules/Currency/Dropdown';
import Combobox from './index';

const title = 'Components/Common/Composite/Combobox';
const props = {
  data: [
    { code: 'EUR', name: 'Euro' },
    { code: 'THB', name: 'Thai Baht' },
    { code: 'USD', name: 'United States Dollar' }
  ],
  display: suggestion => suggestion.name,
  dropdown: ({ data, query }) => (
    <div>
      {data.name} ({data.code})
    </div>
  ),
  id: 'combobox-001',
  input: React.forwardRef(({ onClear, value, ...inputProps }, ref) => (
    <input
      {...inputProps}
      className="form-control"
      ref={ref}
      type="text"
      value={value}
    />
  )),
  onReset: action('onReset'),
  onSelect: action('onSelect'),
  placeholder: 'Combobox'
};

storiesOf(title, module)
  .add('default', () => <Combobox {...props} />, {
    notes: 'Try with EUR, THB, or USD'
  })
  .add('initial value', () => (
    <Combobox {...props} initialValue="Initial value" />
  ))
  .add('override', () => (
    <Combobox {...props} override value="Overrided value" />
  ));
