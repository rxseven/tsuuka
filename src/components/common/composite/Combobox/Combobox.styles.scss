.react-autosuggest__container {
  position: relative;

  &--open {
    input {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }
}

.react-autosuggest__suggestion {
  cursor: pointer;
  font-size: 0.875rem;
  padding: 0.375rem 0.75rem;

  &--highlighted {
    background-color: #cbbde2;
  }
}

.react-autosuggest__suggestions-container {
  display: none;

  &--open {
    background-color: #fff;
    border: 1px solid #beabda;
    border-bottom-left-radius: 0.25rem;
    border-bottom-right-radius: 0.25rem;
    border-top: none;
    display: block;
    max-height: 300px;
    overflow-y: scroll;
    position: absolute;
    width: 100%;
    z-index: 2;
  }
}

.react-autosuggest__suggestions-list {
  list-style-type: none;
  margin: 0;
  padding: 0;
}
