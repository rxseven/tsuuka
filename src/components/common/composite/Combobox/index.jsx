import memoize from 'memoize-one';
import PropTypes from 'prop-types';
import React from 'react';
import Autosuggest from 'react-autosuggest';

import './Combobox.styles.scss';

// Regex
function escapeCharacters(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

// Filter list
// Re-run the filter whenever the list array, keys or filter text changes
const filterList = memoize((list, keys, term) => {
  return list.filter(item =>
    keys.some(key =>
      String(item[key])
        .toLowerCase()
        .includes(term.toLowerCase())
    )
  );
});

// Calculate suggestions for any given input value
function getSuggestions(data, value) {
  // Format input value
  const escapedValue = escapeCharacters(value.trim());

  // Return an empty list if the input value is not defined
  if (escapedValue === '') {
    return [];
  }

  // Return matched suggestions
  // Calculate the latest filtered list. If these arguments haven't changed
  // since the last render, memoized function will reuse the last return value.
  return filterList(data, ['code', 'name'], escapedValue);
}

const propTypes = {
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  display: PropTypes.func.isRequired,
  dropdown: PropTypes.instanceOf(Function).isRequired,
  id: PropTypes.string.isRequired,
  initialValue: PropTypes.string,
  input: PropTypes.instanceOf(Object).isRequired,
  onReset: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  override: PropTypes.bool,
  placeholder: PropTypes.string.isRequired,
  value: PropTypes.string
};

const defaultProps = {
  initialValue: '',
  override: false,
  value: ''
};

// Component
class Combobox extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      suggestions: [],
      value: props.initialValue
    };
  }

  componentDidUpdate() {
    const { value } = this.state;

    // Reset form is neccessary
    this.handleReset(value);
  }

  static getDerivedStateFromProps(props, state) {
    // Force updating state controlled by the parent component
    if (props.override) {
      return { value: props.value };
    }

    return null;
  }

  // Render suggestion list (dropdown)
  // Define how suggestions are rendered
  renderSuggestion = (suggestion, { query }) => {
    const { dropdown: Dropdown } = this.props;

    return <Dropdown data={suggestion} query={query} />;
  };

  // Render input component
  renderInput = props => {
    const { input: Input } = this.props;

    return <Input {...props} />;
  };

  // Update input value and reset form is neccessary
  handleChange = (event, { newValue }) => {
    this.setState({ value: newValue }, () => this.handleReset(newValue));
  };

  // Clear input value and reset form is neccessary
  handleClear = () => {
    this.setState({ value: '' }, () => this.handleReset());
  };

  // Reset form
  handleReset = (value = '') => {
    if (!value) {
      const { onReset } = this.props;
      onReset();
    }
  };

  // Set a selection when suggestion is selected via mouse or keyboard
  handleSuggestionSelected = (event, { suggestion, suggestionValue }) => {
    const { onSelect } = this.props;

    onSelect(suggestion);
  };

  // Update suggestion list
  // This function will be called every time, it might need to update suggestions
  handleSuggestionsFetchRequested = ({ value }) => {
    const { data } = this.props;

    this.setState({ suggestions: getSuggestions(data, value) });
  };

  // Clear suggestion list
  // This function will be called every time, it needs to clear suggestions
  handleSuggestionsClearRequested = () => {
    this.setState({ suggestions: [] });
  };

  // Display suggestions only when input value is at least 2 characters long
  // By default, suggestions are rendered when the input isn't blank
  handleRenderSuggestions = value => value.trim().length > 1;

  render() {
    const { display, id, placeholder } = this.props;
    const { suggestions, value } = this.state;

    const inputProps = {
      onChange: this.handleChange,
      onClear: this.handleClear,
      id,
      placeholder,
      value
    };

    return (
      <Autosuggest
        focusInputOnSuggestionClick={false}
        getSuggestionValue={display}
        highlightFirstSuggestion={false}
        id={id}
        inputProps={inputProps}
        onSuggestionSelected={this.handleSuggestionSelected}
        onSuggestionsClearRequested={this.handleSuggestionsClearRequested}
        onSuggestionsFetchRequested={this.handleSuggestionsFetchRequested}
        renderInputComponent={this.renderInput}
        renderSuggestion={this.renderSuggestion}
        shouldRenderSuggestions={this.handleRenderSuggestions}
        suggestions={suggestions}
      />
    );
  }
}

Combobox.propTypes = propTypes;
Combobox.defaultProps = defaultProps;

export { filterList, getSuggestions };
export default Combobox;
