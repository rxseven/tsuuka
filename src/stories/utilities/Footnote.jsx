import React from 'react';
import styled from 'styled-components';

const Frame = styled.div`
  bottom: 2rem;
  font-size: 14px;
  position: fixed;
`;

function Footnote({ children }) {
  return <Frame>{children}</Frame>;
}

export default Footnote;
