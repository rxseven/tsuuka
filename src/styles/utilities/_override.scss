// Reset button's default styles
button:focus {
  box-shadow: none !important;
  outline: 0 !important;
}
