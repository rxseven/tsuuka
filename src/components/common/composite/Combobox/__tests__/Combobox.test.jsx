import React from 'react';
// import { fireEvent, waitForElement } from 'react-testing-library';

import { factory } from 'tests/utilities';
import Combobox, { filterList, getSuggestions } from '../index';

// Arrange
const source = {
  data: [
    { code: 'EUR', name: 'Euro' },
    { code: 'THB', name: 'Thai Baht' },
    { code: 'USD', name: 'United States Dollar' }
  ],
  display: suggestion => suggestion.name,
  dropdown: ({ data, query }) => (
    <div data-testid="dropdown">
      {data.name} ({data.code})
    </div>
  ),
  id: 'combobox-001',
  input: React.forwardRef(({ onClear, value, ...inputProps }, ref) => (
    <input
      {...inputProps}
      className="form-control"
      ref={ref}
      placeholder="Enter currency"
      type="text"
      value={value}
    />
  )),
  onReset: jest.fn(),
  onSelect: jest.fn(),
  placeholder: 'Combobox'
};

// Setup
function setup(props) {
  return factory(Combobox, source, props);
}

// Test suites
describe('<Combobox />', () => {
  it('should render without crashing', () => {
    setup();
  });

  it('should render an autocomplete field', () => {
    const { queryByPlaceholderText } = setup();

    expect(queryByPlaceholderText('Enter currency')).toBeInTheDocument();
  });

  it('should render initial input value', () => {
    const props = { initialValue: 'Thai Baht' };
    const expected = { value: props.initialValue };
    const { getByPlaceholderText } = setup(props);

    expect(getByPlaceholderText('Enter currency').value).toBe(expected.value);
  });

  it('should render input value from prop when overriding', () => {
    const props = { override: true, value: 'Thai Baht' };
    const expected = { value: props.value };
    const { getByPlaceholderText } = setup(props);

    expect(getByPlaceholderText('Enter currency').value).toBe(expected.value);
  });

  describe('filterList()', () => {
    it('should return filtered list', () => {
      const list = source.data;
      const key = ['code', 'name'];
      const term = 'Thai';
      const expected = { filtered: [list[1]] };

      const result = filterList(list, key, term);

      expect(result).toEqual(expected.filtered);
    });

    it('should return empty list if not found', () => {
      const list = source.data;
      const key = ['code', 'name'];
      const term = 'Japanese Yen';
      const expected = { filtered: [] };

      const result = filterList(list, key, term);

      expect(result).toEqual(expected.filtered);
    });
  });

  describe('getSuggestions()', () => {
    it('should return suggestions', () => {
      const list = source.data;
      const value = 'Thai';
      const expected = { suggestions: [list[1]] };

      const result = getSuggestions(list, value);

      expect(result).toEqual(expected.suggestions);
    });

    it('should return empty list if not found', () => {
      const list = source.data;
      const value = 'Japanese Yen';
      const expected = { suggestions: [] };

      const result = getSuggestions(list, value);

      expect(result).toEqual(expected.suggestions);
    });

    it('should remove whitespace from both sides of the search term', () => {
      const list = source.data;
      const value = ' Thai ';
      const expected = { suggestions: [list[1]] };

      const result = getSuggestions(list, value);

      expect(result).toEqual(expected.suggestions);
    });
  });
});
