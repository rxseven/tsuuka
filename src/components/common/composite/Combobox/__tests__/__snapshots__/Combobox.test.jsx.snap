// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Combobox /> Snapshot tests should render correctly 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="react-autowhatever-combobox-001"
  className="react-autosuggest__container"
  role="combobox"
>
  <input
    aria-activedescendant={null}
    aria-autocomplete="list"
    aria-controls="react-autowhatever-combobox-001"
    autoComplete="off"
    className="form-control"
    id="combobox-001"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="Combobox"
    type="text"
    value=""
  />
  <div
    className="react-autosuggest__suggestions-container"
    id="react-autowhatever-combobox-001"
    role="listbox"
  />
</div>
`;
